File: 2021/0123-best-time-to-buy-and-sell-stock-iii.py
# 链接：https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-iii/solution/yi-kan-jiu-dong-de-jian-dan-ti-jie-by-ed-0o3l/
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        # 5种状态
        dp0 = 0             # 一直不买
        dp1 = - prices[0]   # 到最后也只买入了一笔
        dp2 = float("-inf") # 到最后买入一笔，卖出一笔
        dp3 = float("-inf") # 到最后买入两笔，卖出一笔
        dp4 = float("-inf") # 到最后买入两笔，卖出两笔

        for i in range(1, len(prices)):
            dp1 = max(dp1, dp0 - prices[i])
            dp2 = max(dp2, dp1 + prices[i])
            dp3 = max(dp3, dp2 - prices[i])
            dp4 = max(dp4, dp3 + prices[i])

        return max(dp0, dp2, dp4)

File: hot100/0123-best-time-to-buy-and-sell-stock-iii.go
var minInt = ^int(^uint(0) >> 1)

func maxProfit(prices []int) int {
    if len(prices) == 0 {
        return 0
    }

    // 初始化股票买卖情况
    d0 := 0             // 没有任何买入操作
    d1 := -prices[0]    // 买入一支股票
    d2 := 0             // 买入一支，卖出一支
    d3 := -prices[0]    // 买入两支，卖出一支
    d4 := 0             // 买入两支，卖出两支

    for i := 1; i < len(prices); i++ {
        d1 = max(d1, d0-prices[i])
        d2 = max(d2, d1+prices[i])
        d3 = max(d3, d2-prices[i])
        d4 = max(d4, d3+prices[i])
    }

    return max(d0, d1, d2, d3, d4)
}

func max(nums... int) int {
    if len(nums) == 0 {
        return 0
    }

    n := nums[0]
    for i := range nums {
        if n < nums[i] {
            n = nums[i]
        }
    }

    return n
}
